cpp/src/strings/repeat_strings.cu

```diff
@@ -172,22 +199,259 @@ std::unique_ptr<column> repeat_strings(strings_column_view const& input,
+
+  auto [null_mask, null_count] = cudf::detail::valid_if(
+    validities.begin(), validities.end(), thrust::identity<int8_t>{}, stream, mr);
+
+  return std::make_tuple(std::move(offsets_column),
+                         size_bytes,
+                         null_count > 0 ? std::move(null_mask) : rmm::device_buffer{},
+                         null_count);
+}
+
+template <typename Func>
+auto generate_output_strings(Func& fn,
+                             size_type strings_count,
+                             size_type size_bytes,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
+{
+  auto chars_column = create_chars_child_column(0, size_bytes, stream, mr);
+
+  // Note that if the output chars column has zero size, the function fn should not be called to
+  // avoid accidentally overwriting the offsets.
+  if (size_bytes > 0) {
+    fn.d_chars = chars_column->mutable_view().template data<char>();
+    thrust::for_each_n(
+      rmm::exec_policy(stream), thrust::make_counting_iterator<size_type>(0), strings_count, fn);
+  }
+
+  return chars_column;
+}
+
+/**
+ * @brief Functor to compute string sizes and repeat the input strings, each string is repeated by a
+ * separate number of times.
+ */
+template <class IntType>
+struct compute_size_and_repeat_separately_fn {
+  column_device_view const strings_dv;
+  column_device_view const repeat_times_dv;
+  bool const strings_has_nulls;
+  bool const rtimes_has_nulls;
+
+  offset_type* d_offsets{nullptr};
+
+  // If d_chars == nullptr: only compute sizes of the output strings.
+  // If d_chars != nullptr: only repeat strings.
+  char* d_chars{nullptr};
+
+  // We may need to set `1` or `0` for the validities of the output strings,
+  // and only do that when both input columns have nulls.
+  int8_t* d_validities{nullptr};
+
+  __device__ void operator()(size_type const idx) const noexcept
+  {
+    auto const string_is_valid = !strings_has_nulls || strings_dv.is_valid_nocheck(idx);
+    auto const rtimes_is_valid = !rtimes_has_nulls || repeat_times_dv.is_valid_nocheck(idx);
+
+    // Any null input (either string or repeat_times value) will result in a null output.
+    auto const is_valid = string_is_valid && rtimes_is_valid;
+
+    // When the input string is null, `repeat_times` is also set to 0.
+    // This makes sure that if `repeat_times > 0` then we will always have a valid input string,
+    // and if `repeat_times <= 0` we will never copy anything to the output.
+    auto const repeat_times = is_valid ? repeat_times_dv.element<IntType>(idx) : IntType{0};
+
+    if (!d_chars) {
+      d_offsets[idx] = [&] {
+        if (repeat_times <= 0) { return size_type{0}; }
+        auto const size = static_cast<int64_t>(strings_dv.element<string_view>(idx).size_bytes()) *
+                          static_cast<int64_t>(repeat_times);
+        return size > max_offset ? invalid_size : static_cast<size_type>(size);
```

Comment:
Seems this is checking for overflow when it shouldn't be.

Reply:
The idea is that, not only the sum sizes of strings can cause overflow but each individual string can cause overflow too. If we don't check right at the size computation step, we may miss it as overflow will be wrapped around back to be a valid number.